### ISSUE TITLE
Break up KeyExchange trait into smaller components

### ISSUE THREAD

**Issue by kevinlewi** [novifinancial/opaque-ke#23]
Instead of having one KeyExchange trait which implements generate_ke1/ke2/ke3, we should create traits for KE1 / KE2 / KE3 which implement those functions.

Also can make these new traits generic with respect to the keypair type.

Continuation of #3

**Comment by huitseeker** [comment]
`KeyExchange` has a number of issues:

## Incoherent use of generics

`KeyExchange` has generic methods that take any `KeyPair` format, but it has "at rest" data formats (`KE1State`, `KE1Message`, `KE2Message`) that can't.

Concretely, the fact that the implementations of `KeyExchange` are made to take in any key sort, but can't store them means they either :
1. store key data as a serialized bytes representation, that does not carry its type, or, equivalently, rely on undifferentiated type `Key` (a wrapper around `Vec<u8>` as they do now with constraints `Repr = Key`¹ 
2. or store the *type constructors* of `KE1{State, Message}` & `KE2Message` when Rust adopts generic associated types (RFC 1598).

¹: https://gist.github.com/1a0e6835225521c154cfab81d5a027ed

The problem of 1. (i.e. what we have right now) is that the data is essentially untyped: data (`KE1Message`) serialized by an instance of `KeyExchange` working on XEdDSA can be deserialized and misinterpreted by an instance of `KeyExchange` working on X25519.

The problem of 2. is that it requires unimplemented / nightly features, and is wielding a bazooka to kill a mosquito.

Another approach is to refactor `KeyExchange` so that one of two things happen:
1. the trait `KeyExchange` is generic in a `KeyFormat` parameter itself, and its methods receive this type parameter rather than a  generic parameter local to that method.
2. the trait `KeyExchange` is not generic in its `KeyPair` instance and requires the implementers to concretize their key format.

The  advantage of 1. compared to 2. is that 1. can represent generic approaches to key exchange,  such as TripleDH. 2. requires reimplementations when the KeyPair changes.

## Compositionality: allowing distinct server / client approaches

Currently, somebody that want to customize part of the approach for the library has to reimplement client and server approaches while maybe only wanting to customize one of the two.

## Compositionality: number of messages

Currently, the AKE protocol is a 3-message which performs an authentication (monikered 'finalization') on the last message. Should an implementer want to generate a 4-message protocol (as in the Sigma-R), there is way to extend the trait to cover this, since a new protocol message does not build on the previous ones, and entire new trait is necessary.

**Comment by huitseeker** [comment]
Edit: The "Incoherent use of generics" section above is fixed after #50, #51, #52

**Comment by kevinlewi** [comment]
Thank you for documenting this so far!

**Comment by kevinlewi** [comment]
Closing as this is no longer relevant
